Epona/include/Epona/QuickhullConvexHull.hpp

```diff
@@ -25,512 +25,395 @@
+            indices.push_back(v);
+    std::sort(hull.indices.begin(), hull.indices.end(), std::greater<uint32_t>());
+
+    uint32_t totalOutliersCount = 0;
+    ::FaceOutliers faceOutliers;
+    std::vector<uint32_t> visibleFaces;
+    std::vector<::Ridge> boundary;
+    std::vector<uint32_t> temporaryOutliers;
+
+    do {
+        glm::vec3 mean(0);
+        for (auto i : hull.indices)
+            mean += vertices[i];
+        mean /= hull.indices.size();
+
+        //Find outliers and extreme vertices
+        uint32_t newOutliersCount = 0;
+        faceOutliers.outliers.resize(hull.faces.size());
+        faceOutliers.extremeIndices.resize(hull.faces.size(), UINT32_MAX);
+        faceOutliers.extremeSignedDistances.resize(hull.faces.size(), -FLT_MAX);
+        for (uint32_t v : indices)
         {
-            faceIteratorStack.push_back(faceIt);
+            for (uint32_t f = 0; f < hull.faces.size(); ++f)
+            {
+                float const signedDistance = hull.planes[f].SignedDistance(vertices[v]);
+                if (signedDistance > distanceThreshold) {
+                    faceOutliers.outliers[f].push_back(v);
+                    newOutliersCount++;
+                    if (signedDistance > faceOutliers.extremeSignedDistances[f]) {
+                        faceOutliers.extremeSignedDistances[f] = signedDistance;
+                        faceOutliers.extremeIndices[f] = v;
+                    }
+                    break;
+                }
+            }
         }
 
-        //Find initial visible face
-        typename Faces::iterator visibleFaceIterator = m_faces.end();
-        for (typename Faces::iterator faceIt : faceIteratorStack)
+        //Find visible faces from the first extreme outlier
+        uint32_t extremeOutlierIndex = UINT32_MAX;
+        visibleFaces.clear();
+        visibleFaces.reserve(hull.faces.size() / 2);
+        for (uint32_t f = 0; f < hull.faces.size(); ++f)
         {
-            if (faceIt->GetHyperPlane().SignedDistance(m_vertexBuffer[index]) > 0)
+            if (!faceOutliers.outliers[f].empty())
             {
-                visibleFaceIterator = faceIt;
+                visibleFaces.push_back(f);
+                std::vector<uint32_t> visitedFaces(1, f);
+                visitedFaces.reserve(hull.faces.size() / 2);
+                glm::vec3 const extremeVertex = vertices[faceOutliers.extremeIndices[f]];
+
+                for (uint32_t vf = 0; vf < visibleFaces.size(); ++vf)
+                {
+                    auto adjFaceIt = hull.heds.GetFace(hull.faces[visibleFaces[vf]])->GetAdjacentFaceIterator();
+
+                    for (uint8_t i = 0; i < 3; ++i, ++adjFaceIt) {
+                        //NOTE: will work faster then open address map till ~ visitedFaces.size() < 100
+                        if (std::find(visitedFaces.begin(), visitedFaces.end(), adjFaceIt->index) == visitedFaces.end())
```

Comment:
`std::none_of` instead?

Reply:
@dsodintsov If it would work with value type and not just unary predicate, I would use it. But as it is since C++11, the predicate itself would require a lamba function or some function defined somewhere, which kind of contradicts the reason to use it in a first place.